src/expr/src/linear.rs

```diff
@@ -533,4 +546,60 @@ impl MapFilterProject {
         // This should probably resemble existing scalar cse.
         unimplemented!()
     }
+
+    /// Removes unused expressions from `self.expressions`.
+    ///
+    /// Expressions are "used" if they are relied upon by any output columns
+    /// or any predicates, even transitively. Any expressions that are not
+    /// relied upon in this way can be discarded.
+    pub fn remove_undemanded(&mut self) {
+        // Determine the demanded expressions to remove irrelevant ones.
+        let mut demand = std::collections::HashSet::new();
+        for (_index, pred) in self.predicates.iter() {
+            demand.extend(pred.support());
+        }
+        // Start from the output columns as presumed demanded.
+        // If this is not the case, the caller should project some away.
+        demand.extend(self.projection.iter().cloned());
+        // Proceed in *reverse* order, as expressions may depend on other
+        // expressions that precede them.
+        for index in (0..self.expressions.len()).rev() {
+            if demand.contains(&(self.input_arity + index)) {
+                demand.extend(self.expressions[index].support());
+            }
+        }
+
+        // Maintain a map from initial column identifiers to locations
+        // once we have removed undemanded expressions.
+        let mut remap = HashMap::new();
+        // This map only needs to map elements of `demand` to a new location,
+        // but the logic is easier if we include all input columns (as the
+        // new position is then determined by the size of the map).
+        for index in 0..self.input_arity {
+            remap.insert(index, index);
+        }
+        // Retain demanded expressions, and record their new locations.
+        let mut new_expressions = Vec::new();
+        for (index, expr) in self.expressions.drain(..).enumerate() {
+            if demand.contains(&(index + self.input_arity)) {
+                remap.insert(index + self.input_arity, remap.len());
+                new_expressions.push(expr);
+            }
+        }
+        self.expressions = new_expressions;
+
+        // Update column identifiers.
+        for expr in self.expressions.iter_mut() {
+            expr.permute_map(&remap);
+        }
+        for (pos, pred) in self.predicates.iter_mut() {
+            pred.permute_map(&remap);
+            // Insert predicate as eagerly as it can be evaluated:
+            // just after the largest column in its support is formed.
+            *pos = pred.support().into_iter().max().map(|c| c + 1).unwrap_or(0);
```

Comment:
do we need to resort by `pos` now? its hard for me to track down whether we should have to or not - I _think_ no but its hard to prove

Reply:
Oh, probably! That is a great point. Here I thought I was clever by just recomputing the thing you asked for. :)

Reply:
A different take would be that we should deconstruct into `(map, pred, proj)` and just reconstruct the MFP from that, relying on one code path in the builder to produce a sane representation. Lemme try that out.